src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -4080,6 +4086,18 @@ WorkerSessionFailed(WorkerSession *session)
 
 		PlacementExecutionDone(placementExecution, succeeded);
 	}
+
+	/*
+	 * We might hit the connection timeout. In that case, we should not
+	 * let the connection establishment to continue because the execution
+	 * logic pretends that failed sessions are not going to be used anymore.
+	 *
+	 * That's why we forcefully terminate the connection and mark it as failed
+	 * to trigger the state changes in the executor.
+	 */
+	MultiConnection *connection = session->connection;
+	CitusPQFinish(connection);
```

Comment:
We also do a ShutdownConnection in ConnectionStateMachine. Do we still need both?

Reply:
`ShutdownConnection ` will become no-op, should we care about that?

Reply:
ah, then calling  `CitusPQFinish ` on `WorkerSessionFailed ` breaks lots of test  outputs because we can no longer get the error details to print in `MULTI_CONNECTION_FAILED`. 

One another option could be to move this logic to  `CheckConnectionTimeout()`. Let me see how that looks

Reply:
In the last commit, I restricted the changes to only the case that we are interested. To me, this seems like a missing logic from the initial executor implementation, not a band-aid for the bug.

Reply:
For example, we already have the equivalent function in  `FinishConnectionListEstablishment()` named `CloseNotReadyMultiConnectionStates()`